### PR TITLE
New process for revising submissions, and removal of duplicate participant data

### LINKED ISSUE #230: model: remove participants from HEPSubmission object
There is a `SubmissionParticipant` object defined in `hepdata/modules/permissions/models.py`.  The `HEPSubmission` object in `hepdata/modules/submission/models.py` also defines a `participants` column which links to the `SubmissionParticipant` object:
https://github.com/HEPData/hepdata/blob/2f6526dd2b33bf0f5700bd206dcfb97d2bc81695/hepdata/modules/submission/models.py#L32-L38
https://github.com/HEPData/hepdata/blob/2f6526dd2b33bf0f5700bd206dcfb97d2bc81695/hepdata/modules/submission/models.py#L90-L92
The `SubmissionParticipant` is independent of a version, but `HEPSubmission` depends on a version and the `participants` are not copied when creating a new version:
https://github.com/HEPData/hepdata/blob/2f6526dd2b33bf0f5700bd206dcfb97d2bc81695/hepdata/modules/records/api.py#L355-L359
The code queries both the `SubmissionParticipant` table directly and the `participants` column of the `HEPSubmission` table (for example, for the Dashboard display) leading to some inconsistencies for versions of 2 or greater.  It would be simpler to only use the `SubmissionParticipant` table throughout the code and remove the `participants` column of the `HEPSubmission` table.  After the code is modified to remove use of the `participants` column, perhaps the DB cleanup could be done at the same time as #139 (?).

### LINKED ISSUE #231: records: change behaviour of "Revise Submission" button so upload not needed at the same time
Currently, either Coordinators or Uploaders can create a new version by clicking the "Revise Submission" on the record.  However, they need to make an upload at the same time via the web interface, and upload is not possible via the CLI method for a finalised record.  I would prefer that only a Coordinator can click "Revise Submission" which creates a new version (a new `HEPSubmission` object with `todo` status), then the upload is done in a separate step (by either the Coordinator or an Uploader).  This makes the behaviour more similar for version 1 compared to later versions and it means that the Uploader can make the initial upload for version 2 (or later) via the CLI method if they prefer rather than relying on web upload.

### PR DESCRIPTION
**Pull request by alisonrclarke**
Fixes #230, #231

After the code has been deployed, the`submission_participant_link` table should be dropped from the DB:

```sql
DROP TABLE submission_participant_link;
```

### CODE REVIEW

**Review by GraemeWatt — CHANGES_REQUESTED** [review]
I've had a first look at testing the new revision process.  The main problem I see is that when I click "Revise Submission" in the new widget it has no effect.  I get an error message in the JavaScript console: `TypeError: HEPDATA.hepdata_record.revise_submission is not a function. (In 'HEPDATA.hepdata_record.revise_submission()', 'HEPDATA.hepdata_record.revise_submission' is undefined)`.  I also don't see any Uploaders or Reviewers listed for records where there should be some, and loading the widget gives a similar error message in the JavaScript console `TypeError: HEPDATA.hepdata_record.load_revise_submission is not a function.`.

Most of the content of the "Revise Submission" widget can be suppressed for records with no Uploaders/Reviewers (for imported/migrated records).  In this case, there should just be a message like "There are no current Uploaders/Reviewers.  You can add new ones from your Dashboard after the new version is created.".  Finally, it would look better if the text "Optional additional message:" was either centred or appeared at the top of the input text box to the right instead of being displayed at the bottom.

**Comment by alisonrclarke** [comment]
Did you rebuild the javascript (`hepdata webpack buildall`) before testing? (Sorry, should have mentioned that in the PR description.)

**Comment by GraemeWatt** [comment]
> Did you rebuild the javascript (`hepdata webpack buildall`) before testing? (Sorry, should have mentioned that in the PR description.)

I had run `./scripts/clean_assets.sh`, but I've checked that `hepdata webpack buildall` doesn't help.  I just tried removing the `$HOME/.virtualenvs/hepdata/var/hepdata-instance/static/dist` directory before running `./scripts/clean_assets.sh` again.  Now I get an error `FileNotFoundError: [Errno 2] No such file or directory: '$HOME/.virtualenvs/hepdata/var/hepdata-instance/static/dist/manifest.json'` as in the failed CI jobs.  Maybe it worked before because I hadn't installed into a fresh virtualenv after the move to Webpack.

**Comment by alisonrclarke** [comment]
> I had run `./scripts/clean_assets.sh`, but I've checked that `hepdata webpack buildall` doesn't help. I just tried removing the `$HOME/.virtualenvs/hepdata/var/hepdata-instance/static/dist` directory before running `./scripts/clean_assets.sh` again. Now I get an error `FileNotFoundError: [Errno 2] No such file or directory: '$HOME/.virtualenvs/hepdata/var/hepdata-instance/static/dist/manifest.json'` as in the failed CI jobs. Maybe it worked before because I hadn't installed into a fresh virtualenv after the move to Webpack.

I haven't got to the bottom of what's changed that means the webpack build isn't creating `manifest.json` in the `static/dist` directory any more (it was previously doing so or the CI would never have worked), but there's a fix in [this commit](https://github.com/HEPData/hepdata/commit/ed09ab7d35b2aee412f4b479dced218394346ee0) that you could try to get it working locally.

**Comment by GraemeWatt** [comment]
Thanks for the fix.  Now it looks like we can instead bump `invenio-assets` after they fixed this bug [yesterday](https://github.com/inveniosoftware/invenio-assets):

Version 1.2.6 (released 2020-05-18)

* Fixes missing "manifest.json". The issue was caused by the  webpack-bundle-tracker changing behavior of where to output the manifest.json  file between version v1.0.0-alpha1 and v1.0.0 (released 4 days ago).

Sorry that you spent so much time investigating this problem.  It certainly doesn't look obvious.

**Comment by alisonrclarke** [comment]
> Most of the content of the "Revise Submission" widget can be suppressed for records with no Uploaders/Reviewers (for imported/migrated records). In this case, there should just be a message like "There are no current Uploaders/Reviewers. You can add new ones from your Dashboard after the new version is created.". 

👍 Will make that change now.

> Finally, it would look better if the text "Optional additional message:" was either centred or appeared at the top of the input text box to the right instead of being displayed at the bottom.

I think this is another indication of the webpack build not having run properly (or assets not refreshed in your browser). Here's how it looks for me:

Screenshot 2021-05-19 at 12 07 37

(And you can see it on [this SauceLabs run](https://app.eu-central-1.saucelabs.com/tests/0130092dd3914e1f8e2a4882a3f0c679#21) at 29s in.

**Comment by GraemeWatt** [comment]
Yes, my [previous issues](https://github.com/HEPData/hepdata/pull/337#pullrequestreview-662316689) were solved with the Webpack build, apart from the suggestion to suppress content if there are no Uploaders/Reviewers.

Another request: can the text of the invite email (`invite.txt`) please include the email address of the Coordinator sending the invitation, i.e. after "`You have been requested `" insert "`by {{ coordinator_email }} `".  Sometimes it's not clear to the Uploader/Reviewer who made the request, especially for a revision made some time after the original submission.

**Comment by GraemeWatt** [comment]
The changes look good, thanks.  I made some minor tweaks after finding some issues when testing:

- The value of the `invite_token` given in the CLI command in the invitation email was `None`, because it was only being set if permissions had not yet been claimed (i.e. empty `user_account`).  I changed the code to pass both `invite_token` and `user_account` to `invite.html` to separate the previous dual use of `invite_token`.
- The content of the "Revise Submission" widget should be suppressed if there is no primary Uploader.  The existence of primary Reviewers and secondary Uploaders/Reviewers are not so relevant.
- There could possibly be multiple primary Uploaders for the original submission who should all receive an email invitation for the revision, not just one.
- I made some cosmetic changes to wording like capitalising the role names (Uploader/Reviewer/Coordinator) and fixing a grammatical error ("a Uploader" → "an Uploader") that's been annoying me for years.

I think this PR is ready to merge if you're happy with it, @alisonrclarke.

(The CI initially failed for the last commit because [Zenodo](https://zenodo.org) was down for an hour or so.  I hadn't realised that the HEPData site depends on a [badge](https://zenodo.org/badge/DOI/10.5281/zenodo.197109.svg) from Zenodo, even though it was me who added the link in 2016, probably because it is rare that Zenodo goes down.)

File: hepdata/modules/records/api.py [hunk 1]
```diff
@@ -392,10 +392,10 @@ def create_new_version(recid, user, notify_uploader=True, uploader_message=None)
...
 
         if notify_uploader:
-            uploader = SubmissionParticipant.query.filter_by(
+            uploaders = SubmissionParticipant.query.filter_by(
                 role='uploader', publication_recid=recid, status='primary'
-                ).first()
-            if uploader:
+                )
+            for uploader in uploaders:
                 record_information = get_record_by_id(recid)
```
Comment (alisonrclarke) [inline hepdata/modules/records/api.py]: Worth moving this line outside of the loop?

Comment (alisonrclarke) [reply hepdata/modules/records/api.py]: Looks good to me.

**Review by GraemeWatt — APPROVED** [review]

**Pull request merged**